Uncaught C exceptions

```
2024-07-05 03:23:14.990 - amici.swig_wrappers - WARNING - [cEFM192A__EGF+cEFM192A][AMICI:Inf] AMICI encountered a Inf value for sxdot[0] (0) at t=0.000000 
2024-07-05 03:23:14.990 - amici.swig_wrappers - WARNING - [cEFM192A__EGF+cEFM192A][AMICI:Inf] AMICI encountered a Inf value for sxdot[0] (0) at t=0.000000 
2024-07-05 03:23:14.990 - amici.swig_wrappers - WARNING - [cEFM192A__EGF+cEFM192A][AMICI:Inf] AMICI encountered a Inf value for sxdot[0] (0) at t=0.000000 
2024-07-05 03:23:14.990 - amici.swig_wrappers - WARNING - [cEFM192A__EGF+cEFM192A][AMICI:Inf] AMICI encountered a Inf value for sxdot[0] (0) at t=0.000000 
2024-07-05 03:23:14.990 - amici.swig_wrappers - WARNING - [cEFM192A__EGF+cEFM192A][AMICI:Inf] AMICI encountered a Inf value for sxdot[0] (0) at t=0.000000 
2024-07-05 03:23:14.990 - amici.swig_wrappers - WARNING - [cEFM192A__EGF+cEFM192A][AMICI:Inf] AMICI encountered a Inf value for sxdot[0] (0) at t=0.000000 
2024-07-05 03:23:14.990 - amici.swig_wrappers - WARNING - [cEFM192A__EGF+cEFM192A][AMICI:Inf] AMICI encountered a Inf value for sxdot[0] (0) at t=0.000000 
2024-07-05 03:23:14.991 - amici.swig_wrappers - WARNING - [cEFM192A__EGF+cEFM192A][AMICI:Inf] AMICI encountered a Inf value for sxdot[0] (0) at t=0.000000 
2024-07-05 03:23:14.991 - amici.swig_wrappers - WARNING - [cEFM192A__EGF+cEFM192A][AMICI:Inf] AMICI encountered a Inf value for sxdot[0] (0) at t=0.000000 
2024-07-05 03:23:14.991 - amici.swig_wrappers - WARNING - [cEFM192A__EGF+cEFM192A][AMICI:Inf] AMICI encountered a Inf value for sxdot[0] (0) at t=0.000000 
2024-07-05 03:23:14.991 - amici.swig_wrappers - DEBUG - [cEFM192A__EGF+cEFM192A][CVODES:CVode:OTHER] AMICI ERROR: in module CVODES in function CVode : At t = 0 repeated recoverable sensitivity right-hand side function errors. 
2024-07-05 03:23:14.991 - amici.swig_wrappers - ERROR - [cEFM192A__EGF+cEFM192A][OTHER] AMICI simulation failed: CVODE routine reInitPostProcess failed with error code -43. tout: 5.5, t: 0.
2024-07-05 03:23:15.032 - amici.swig_wrappers - DEBUG - [cEFM192A__EGF+cEFM192A][CVODES:CVode:OTHER] AMICI ERROR: in module CVODES in function CVode : At t = 0, the setup routine failed in an unrecoverable manner. 
2024-07-05 03:23:15.033 - amici.swig_wrappers - ERROR - [cEFM192A__EGF+cEFM192A][OTHER] AMICI simulation failed: CVODE routine reInitPostProcess failed with error code -6. tout: 5.5, t: 0.
terminate called after throwing an instance of 'amici::CvodeException'
  what():  CVODE routine CVodeGetSens failed with error code -25. 
```
While the underlying issue is Inf in some of the equations, this shouldn't lead to an uncaught c++ exception that terminates python.

> I added a test case in https://github.com/AMICI-dev/AMICI/pull/2484, does this pass for you?

The first version of this PR passed for me. The current version fails with
```
FAILED test_swig_interface.py::test_python_exceptions - RuntimeError: Dimension mismatch. Size of parameters does not match number of model parameters.
```

but I guess this still counts as a pass, since there is no uncaught exception.

> Potential issues:
> 
>     * different swig versions
> 
>     * updating amici without re-installation
> 
>     * re-installing amici without re-importing the model(?)
> 
>     * linking issues

The 2nd and 3rd aren't the case for me I think. My SWIG is 4.0.2. Not sure how to check the linking issue. I'll try to create an MWE then.

--- a/python/tests/test_swig_interface.py
+++ b/python/tests/test_swig_interface.py
@@ -572,6 +572,7 @@ def test_python_exceptions(sbml_example_presimulation_module):
     rdata = amici.runAmiciSimulation(model, solver)
     assert rdata.status == amici.AMICI_FIRST_RHSFUNC_ERR
 
+    edata = amici.ExpData(1, 1, 1, [1])
     rdatas = amici.runAmiciSimulations(
         model, solver, [edata, edata], failfast=True, num_threads=1
     )

--- a/python/tests/test_swig_interface.py
+++ b/python/tests/test_swig_interface.py
@@ -562,11 +562,20 @@ def test_python_exceptions(sbml_example_presimulation_module):
     ):
         amici.runAmiciSimulation(model, solver, edata)
 
+    amici.runAmiciSimulations(
+        model, solver, [edata, edata], failfast=True, num_threads=1
+    )
+
     # model throws, base catches, swig-exception handling is not involved
     model.setParameters([nan] * model.np())
     model.setTimepoints([1])
     rdata = amici.runAmiciSimulation(model, solver)
     assert rdata.status == amici.AMICI_FIRST_RHSFUNC_ERR
+
+    rdatas = amici.runAmiciSimulations(
+        model, solver, [edata, edata], failfast=True, num_threads=1
+    )
+    assert rdatas[0].status == amici.AMICI_FIRST_RHSFUNC_ERR
 
     # model throws, base catches, swig-exception handling is involved
     from amici._amici import runAmiciSimulation

--- a/python/tests/test_swig_interface.py
+++ b/python/tests/test_swig_interface.py
@@ -567,3 +567,12 @@ def test_python_exceptions(sbml_example_presimulation_module):
     model.setTimepoints([1])
     rdata = amici.runAmiciSimulation(model, solver)
     assert rdata.status == amici.AMICI_FIRST_RHSFUNC_ERR
+
+    # model throws, base catches, swig-exception handling is involved
+    from amici._amici import runAmiciSimulation
+
+    with pytest.raises(
+        RuntimeError, match="AMICI failed to integrate the forward problem"
+    ):
+        # rethrow=True
+        runAmiciSimulation(solver, None, model.get(), True)

--- a/python/tests/test_swig_interface.py
+++ b/python/tests/test_swig_interface.py
@@ -5,7 +5,7 @@ Test getters, setters, etc.
 
 import copy
 import numbers
-
+from math import nan
 import pytest
 
 import amici
@@ -534,3 +534,36 @@ def test_rdataview(sbml_example_presimulation_module):
 
     # field names are included by dir()
     assert "x" in dir(rdata)
+
+
+def test_python_exceptions(sbml_example_presimulation_module):
+    """Test that C++ exceptions are correctly caught and re-raised in Python."""
+
+    # amici-base extension throws and its swig-wrapper catches
+    solver = amici.CVodeSolver()
+    with pytest.raises(
+        RuntimeError, match="maxsteps must be a positive number"
+    ):
+        solver.setMaxSteps(-1)
+
+    # model extension throws and its swig-wrapper catches
+    model = sbml_example_presimulation_module.get_model()
+    with pytest.raises(RuntimeError, match="Steadystate mask has wrong size"):
+        model.set_steadystate_mask([1] * model.nx_solver * 2)
+
+    # amici-base extension throws and its swig-wrapper catches
+    edata = amici.ExpData(1, 1, 1, [1])
+    # too short sx0
+    edata.sx0 = (1, 2)
+    with pytest.raises(
+        RuntimeError,
+        match=r"Number of initial conditions sensitivities \(36\) "
+        r"in model does not match ExpData \(2\).",
+    ):
+        amici.runAmiciSimulation(model, solver, edata)
+
+    # model throws, base catches, swig-exception handling is not involved
+    model.setParameters([nan] * model.np())
+    model.setTimepoints([1])
+    rdata = amici.runAmiciSimulation(model, solver)
+    assert rdata.status == amici.AMICI_FIRST_RHSFUNC_ERR

--- a/src/amici.cpp
+++ b/src/amici.cpp
@@ -282,17 +282,28 @@ std::vector<std::unique_ptr<ReturnData>> runAmiciSimulations(
 #pragma omp parallel for num_threads(num_threads)
 #endif
     for (int i = 0; i < (int)edatas.size(); ++i) {
-        auto mySolver = std::unique_ptr<Solver>(solver.clone());
-        auto myModel = std::unique_ptr<Model>(model.clone());
+        // must catch exceptions in parallel section to avoid termination
+        try {
+            auto mySolver = std::unique_ptr<Solver>(solver.clone());
+            auto myModel = std::unique_ptr<Model>(model.clone());
 
-        /* if we fail we need to write empty return datas for the python
-         interface */
-        if (skipThrough) {
-            ConditionContext conditionContext(myModel.get(), edatas[i]);
+            /* if we fail we need to write empty return datas for the python
+             interface */
+            if (skipThrough) {
+                ConditionContext conditionContext(myModel.get(), edatas[i]);
+                results[i]
+                    = std::unique_ptr<ReturnData>(new ReturnData(solver, model)
+                    );
+            } else {
+                results[i] = runAmiciSimulation(*mySolver, edatas[i], *myModel);
+            }
+        } catch (std::exception const& ex) {
             results[i]
                 = std::unique_ptr<ReturnData>(new ReturnData(solver, model));
-        } else {
-            results[i] = runAmiciSimulation(*mySolver, edatas[i], *myModel);
+            results[i]->status = AMICI_ERROR;
+            results[i]->messages.push_back(
+                LogItem(LogSeverity::error, "OTHER", ex.what())
+            );
         }
 
         skipThrough |= failfast && results[i]->status < 0;

